V2.0

FedeClaudi opened the pull request:
Hey guys, 
the draft of `brainrender v2.0` is here.
It has most of the functionality of previous version with 1/4 the number of LOCs (`1041` currently). 
This was achieved by limiting the number of options to do the same thing (e.g. instead of 5 ways to render streamlines now there's 1). Hopefully this makes both the code and the user experience more straightforward.

At this point I'd like to know what you think of the way code is organized now.

Btw, tests coverage are at 90% and tests also run the example files (thanks @vigji )

**main changes**
- The main difference is how code is dived in files. Most things are straightforward (`scene.py`, `atlas.py`...), one thing to notice is that by convention files starting with `_xxx.py` have utility code.
- `scene` has much fewer methods, you can add brain regions, silhouette, labels or use the generic `add` method for the rest
- `Actor` is a new class, every mesh is converted to an `Actor` before rendering
- `actors` includes modules like `points` that have classes and functions to generate `Actor` instances from data, e.g. `Points` makes what was `cells` before
- `Points` now works by taking an `Nx3` numpy array as input instead of dataframes, also currently colors can only be specified as a list of colors (@vigji  maybe we can add a dedicate function to make colors from metadata in `_colors.py`)

as a consequence of these changes, the way stuff is added to the scene will be a bit different, will add working examples soon.

**working functionality**
- [x] Actor
- [x] Scene
- [x] Render
- [x] Points
- [x] Streamlines
- [x] Neuron
- [x] Point
- [x] Cylinder
- [x] allen brain atlas Gene Expression
- [x] screenshot


**missing functionality**
- [x] video making class (is this used???)
- [x] `SHARPTRACK`, I don't know how much this is used.. it's not a lot of code though so might include it for now
- [x] HTML export


**next steps**
- [x] docstring
- [x] add type hints
- [x] **make new examples**
- [x] update docs
- [x] fix failing travis (@adamltyson  might need your help here)
- [x] fix setup/requirements
- [x] integrate with GUI

adamltyson commented:
Hi @FedeClaudi, I'll take some time to look through this properly later, but I think this is really nice. Some random points:
* I would somehow move the `aba` subpackage to indicate that it's not "core" brainrender, i.e. you can't use it with every atlas. Something like `brainrender.atlas_specific.aba`
* I use the video making class, but I only use it for super simple stuff (e.g. rotation)
* IMO, the sharptrack stuff, however simple is a perfect candidate to not be in brainrender, but to be an example script on how to get external data into BR

` The build failed, just like the previous build.` - Travis is so harsh! Happy to take a look at this.

FedeClaudi commented:
Very good points, thanks I'll add that in.

> Travis is so harsh! Happy to take a look at this.

yes please!

FedeClaudi commented:
hey @adamltyson this still fails with `The command "travis_wait 50 pytest" exited with 2.`
On my machine tests take about 150s so not sure what's going on?

adamltyson commented:
It's failing for me locally with the same error:
```
../../miniconda3/envs/brainrender/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_neuron.py:2: in <module>
    from brainrender.actors import Neuron, get_neurons
E   ImportError: cannot import name 'get_neurons' from 'brainrender.actors' (/home/adam/projects/brainrender/brainrender/actors/__init__.py)
```

FedeClaudi commented:
that line should be: `from brainrender.actors import Neuron, make_neurons`, which it is ?
Tests are passing locally for me..

adamltyson commented:
Not sure what was going on, I updated just before I wrote that message. Updated again and all is good. I would hope the next commit will pass.

FedeClaudi commented:
Hey @adamltyson  and @vigji, I've uncluded the GUI into the package.
I've tested it manually at it works fine, but I have no idea how to set up extensive unit tests for it, so for now there's no tests and it's excluded from the coverage report (didn't want to spoil that beautiful 95% coverage), but that's not great.

If you have any idea on how to do this, let me know!

FedeClaudi commented:
hey @vigji when you have a second can you have a look at this? (no rush)

vigji commented on brainrender/actor.py:
```diff
@@ -41,47 +55,62 @@ def __str__(self):
+            Create a new Actor with a sphere and a text
+            labelling this actor
+        """
         labels = make_actor_label(
```
Also, an actor will always be referenced to a scene and its atlas, so probably atlas should just be passed to the actor init

FedeClaudi replied:
>This usage of make_actor_label is a bit funny in that it's functional coding on something that it's clearly object-oriented - why not having it as a method of the class?

It's a relative long (80LOC) and ugly bit of code, so I wanted to hide it away from the class definition. 
I've moved the function to `actor` (from `_actor`) to remove the need for `_actor`, but left it out of the class for now.
I know it's now very elegant but might be okay for now?

> Also, an actor will always be referenced to a scene and its atlas, so probably atlas should just be passed to the actor init

Yeah, that could be done. Though the atlas attribute is only used to create the labels, and even then only to make sure that the labels are in the correct hemisphere... 

This whole label's business is quite ugly for both handling the creation and the rendering of the labels...

vigji replied:
I do agree that this bit of code, having the atlas just for the labels creation, looks a bit cumbersome...I will see if I can come out with more complete suggestions on how this could be improved

FedeClaudi replied:
okay thanks!

FedeClaudi replied:
I'd be tempted to scrap the whole thing to be honest

vigji commented on brainrender/actors/neurons.py:
```diff
@@ -0,0 +1,93 @@
+            Creates an Actor representing a single neuron's morphology
+            :param neuron: path to .swc file, Mesh, Actor or Neuron from morphapi.morphology
+            :param alpha: float
+            :param color: str,
```
as suggested above, I  think that support for RGB is convenient as it is what most users would use by default to customise stuff

vigji commented on brainrender/actors/streamlines.py:
```diff
@@ -0,0 +1,154 @@
+import pandas as pd
```
I would restructure this module to keep here (under `brainrender.actors`) the part that deals with the graphics, and move to the `atlas_specific` (but see relative note) the database part

FedeClaudi replied:
fair enough, I've done so.

vigji commented on brainrender/atlas.py:
```diff
@@ -0,0 +1,135 @@
+            self, atlas_name=atlas_name, print_authors=False
+        )
+
+    def get(self, _type, *args, **kwargs):
```
this getter looks a bit funny to me, what's its purpose? Sounds a bit too general to be useful, I guess that stuff will always be like get_plane or get_region

FedeClaudi replied:
yep, I first put it in thinking that more stuff could be `got` from the atlas but it is now redundant, removing.

vigji commented on brainrender/atlas_specific/aba_gene_expression/api.py:
```diff
@@ -93,9 +97,13 @@ def get_gene_experiments(self, gene_symbol):
             if isinstance(gene_symbol, int):  # it's an ID, get symbol
                 gene_symbol = self.get_gene_symbol_by_id(gene_symbol)
                 if gene_symbol is None:
-                    raise ValueError("Invalid gene_symbol argument")
```
Is there a reason to handle this complicated nested statements here and not just failing in the method get_gene_symbol?

FedeClaudi replied:
nope, removed. thanks

vigji commented on brainrender/atlas_specific/aba_gene_expression/api.py:
```diff
@@ -118,8 +126,8 @@ def download_gene_data(self, gene):
             self.genes = self.get_all_genes()
         if str(gene) not in self.genes.id.values:
-            raise ValueError(
-                f"The gened {gene} is not in the list of available genes"
```
Is it actually so common to get so many misuses of this class to justify the amount of custom errors you're throwing? What about leaving the upstream failing by itself?

vigji commented on brainrender/camera.py:
```diff
@@ -0,0 +1,102 @@
+from .cameras import cameras
+
+
+def get_camera(camera):
```
It looks like this module and the context of `cameras.py` could be refactored in a nice `Camera` class!

FedeClaudi replied:
I prefer to leave it as it is for now, but I will think about it. 

Almost always users will want to either pass a camera name or a dictionary of parameters to `scene.render`, everything else is just internal checks to make sure that camera arguments are correct.
If there was the need to expose more camera functionality to the user it would make sense to turn it into a `Camera` class, but as such there wouldn't be much benefit. 

This approach is still quite flexible (e.g. the new `Animation` class interpolates between two cameras) and (I hope) not too confusing for developers.

vigji replied:
even if you don't expose it it could be helpful for reading the code. But yeah, no priority, until you want people to use it

vigji reviewed (commented):
Overall it looks like a neat refactoring! Definitively going in the proper direction.
I left some comments here and there. Also, this was done only by reading through the commits changes, I will probably give more comments when I play around with the code a bit more.
As a minor naming note, I would probably rename atlas_specific module to something more informative, eg dataset_interfaces/api/io or similar. Also, for the stuff that is there you might want to think a bit (only a bit!) about adding abstract classes for the stuff that then you implement for the allen-specific dataset. This would force you to separate what you need on the "frontend" (the brainrender viewer backend) and the backend (the actual allen database); and if eg one day I will need to do something similar for the fish dataset, I just start from the general class where the interactions with the Scene are already established, and I just implement a new backend. Also, connectome stuff should probably end up here with similar considerations

FedeClaudi commented on brainrender/_colors.py:
```diff
@@ -0,0 +1,115 @@
+
+
+# Some nice colors
+red = "#fc330f"
```
most `vedo` functions will accept rgb colors and support matplotlib's color names.
It's not explicitly referenced in the docstrings because hex or color names are easier to handle.

FedeClaudi commented on brainrender/_colors.py:
```diff
@@ -0,0 +1,115 @@
+mocassin = "#FFE4B5"
+
+
+def colorMap(value, name="jet", vmin=None, vmax=None):
```
it was inherited from vedo, fixed

FedeClaudi commented on brainrender/actor.py:
```diff
@@ -4,34 +4,48 @@
 import numpy as np
 
-from brainrender.Utils.scene_utils import make_actor_label
+from ._actor import make_actor_label
```
fixed thanks

FedeClaudi commented:
hey @vigji thanks for taking the time to go through the code, as always it is very much appreciated (and useful).
I've addressed most comments directly and will push updated code soon. 

> adding abstract classes for the stuff that then you implement for the allen-specific dataset.

Honestly, I don't think it's worth it. Both the gene expression and the streamlines rely **heavily** on the allen SDK so I doubt there would be much that can be used for other datasets. 
In terms of taking the data and creating actors from it, that's already in place for the streamlines. Perhaps it could be useful to make a `Volumetric` actor for stuff like the gene expression data, though

vigji commented:
> In terms of taking the data and creating actors from it, that's already in place for the streamlines. Perhaps it could be useful to make a `Volumetric` actor for stuff like the gene expression data, though

I think this would already move it significantly in this direction!

FedeClaudi commented:
Okay, I've added this. 
Btw the build is failing because of: https://github.com/marcomusy/vedo/issues/243 (I think)

FedeClaudi commented:
Hey @vigji  and @adamltyson, the tests are now passing and everything's ready. 
So tomorrow I will update docs.brainrender.org and release v2.0!
If you have any other change to suggest or comment, now is the time!

adamltyson commented:
As this is a fairly major change, it would probably be useful if you make a docs page for the main changes from v1 -> v2.

Otherwise,  I think it's good to go!
